mutable_batch/src/partition.rs

```diff
@@ -0,0 +1,222 @@
+//! Functions for partitioning rows from a [`MutableBatch`]
+//!
+//! The returned ranges can then be used with `MutableBatch::extend_from_range`
+
+use crate::column::{Column, ColumnData};
+use crate::MutableBatch;
+use chrono::{format::StrftimeItems, TimeZone, Utc};
+use data_types::database_rules::{PartitionTemplate, TemplatePart};
+use schema::TIME_COLUMN_NAME;
+use std::ops::Range;
+
+/// Returns an iterator identifying consecutive ranges for a given partition key
+pub fn partition_batch<'a>(
+    batch: &'a MutableBatch,
+    table_name: &'a str,
+    template: &'a PartitionTemplate,
+) -> impl Iterator<Item = (String, Range<usize>)> + 'a {
+    rle_encode(partition_keys(batch, table_name, template))
+}
+
+enum Template<'a> {
+    Table(&'a str),
+    Column(&'a Column, &'a str),
+    TimeFormat(&'a [i64], StrftimeItems<'a>),
+}
+
+impl<'a> Template<'a> {
+    fn fmt<W: std::fmt::Write>(&self, idx: usize, out: &mut W) -> std::fmt::Result {
```

Comment:
```suggestion
    /// Creates partition key for row `idx` according the template.
    fn partition_key<W: std::fmt::Write>(&self, idx: usize, out: &mut W) -> std::fmt::Result {
```

I was wondering what you're formatting here and why this ain't `Display` or `Debug`.